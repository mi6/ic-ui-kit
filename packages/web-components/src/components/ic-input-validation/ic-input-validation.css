ic-input-validation {
  width: var(--input-width, 20rem);
  margin-top: var(--ic-space-xs);
  display: flex;
}

ic-input-validation.fullwidth {
  width: 100%;
}

ic-input-validation span.status-icon {
  padding-right: var(--ic-space-xxs);
}

ic-input-validation span.status-icon > svg {
  height: 1.25rem;
  width: 1.25rem;
}

ic-input-validation span.icon-success > svg {
  fill: var(--ic-status-success);
}

ic-input-validation span.icon-error > svg {
  fill: var(--ic-status-error);
}

ic-input-validation span.icon-warning > svg {
  fill: var(--ic-status-warning-mid);
}

ic-input-validation .statustext {
  flex-grow: 1;
}
