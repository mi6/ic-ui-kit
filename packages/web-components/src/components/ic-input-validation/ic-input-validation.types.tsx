export type IcAriaLiveModeVariants = "polite" | "assertive";
