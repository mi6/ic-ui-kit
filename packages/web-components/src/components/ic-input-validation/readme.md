# ic-input-validation



<!-- Auto Generated Below -->


## Properties

| Property               | Attribute        | Description                                                            | Type                                      | Default     |
| ---------------------- | ---------------- | ---------------------------------------------------------------------- | ----------------------------------------- | ----------- |
| `ariaLiveMode`         | `aria-live-mode` | The ARIA live mode to apply to the message.                            | `"assertive" \| "polite"`                 | `"polite"`  |
| `for`                  | `for`            | The ID of the form element the validation is bound to.                 | `string`                                  | `undefined` |
| `fullWidth`            | `full-width`     | If `true`, the input validation will fill the width of the container.  | `boolean`                                 | `false`     |
| `message` _(required)_ | `message`        | The validation message to display.                                     | `string`                                  | `undefined` |
| `status`               | `status`         | The status of the validation - e.g. 'error' \| 'warning' \| 'success'. | `"" \| "error" \| "success" \| "warning"` | `""`        |


## Slots

| Slot                             | Description                                                    |
| -------------------------------- | -------------------------------------------------------------- |
| `"validation-message-adornment"` | Content will be placed to the right of the validation message. |


## Dependencies

### Used by

 - [ic-checkbox-group](../ic-checkbox-group)
 - [ic-radio-group](../ic-radio-group)
 - [ic-select](../ic-select)
 - [ic-text-field](../ic-text-field)

### Depends on

- [ic-typography](../ic-typography)

### Graph
```mermaid
graph TD;
  ic-input-validation --> ic-typography
  ic-checkbox-group --> ic-input-validation
  ic-radio-group --> ic-input-validation
  ic-select --> ic-input-validation
  ic-text-field --> ic-input-validation
  style ic-input-validation fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


