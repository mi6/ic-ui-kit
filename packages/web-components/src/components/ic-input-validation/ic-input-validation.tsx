import { Element, Component, Host, Prop, h } from "@stencil/core";

import {
  IcInformationStatus,
  IcInformationStatusOrEmpty,
} from "../../utils/types";
import errorIcon from "../../assets/error-icon.svg";
import successIcon from "../../assets/success-icon.svg";
import warningIcon from "../../assets/warning-icon.svg";

import {
  getInputValidationTextID,
  onComponentRequiredPropUndefined,
} from "../../utils/helpers";
import { IcAriaLiveModeVariants } from "./ic-input-validation.types";

const icon = {
  [IcInformationStatus.Warning]: warningIcon,
  [IcInformationStatus.Error]: errorIcon,
  [IcInformationStatus.Success]: successIcon,
};

/**
 * @slot validation-message-adornment - Content will be placed to the right of the validation message.
 */
@Component({
  tag: "ic-input-validation",
  styleUrl: "ic-input-validation.css",
})
export class InputValidation {
  @Element() el: HTMLIcInputValidationElement;

  /**
   *  The ARIA live mode to apply to the message.
   */
  @Prop() ariaLiveMode: IcAriaLiveModeVariants = "polite";

  /**
   * The ID of the form element the validation is bound to.
   */
  @Prop() for: string;

  /**
   *  If `true`, the input validation will fill the width of the container.
   */
  @Prop() fullWidth: boolean = false;

  /**
   * The validation message to display.
   */
  @Prop() message!: string;

  /**
   * The status of the validation - e.g. 'error' | 'warning' | 'success'.
   */
  @Prop() status: IcInformationStatusOrEmpty = "";

  componentDidLoad(): void {
    onComponentRequiredPropUndefined(
      [{ prop: this.message, propName: "message" }],
      "Input Validation"
    );
  }

  render() {
    const displayIcon = this.status !== "" ? icon[this.status] : "";
    const id = getInputValidationTextID(this.for);
    return (
      <Host
        class={{
          [this.status]: this.status !== "",
          ["fullwidth"]: this.fullWidth,
        }}
      >
        {displayIcon !== "" && (
          <span
            class={{
              ["status-icon"]: true,
              [`icon-${this.status}`]: true,
            }}
            innerHTML={displayIcon}
          />
        )}

        <ic-typography variant="caption" class="statustext">
          <span aria-live={this.ariaLiveMode} id={id}>
            {this.message}
          </span>
        </ic-typography>

        <slot name="validation-message-adornment"></slot>
      </Host>
    );
  }
}
