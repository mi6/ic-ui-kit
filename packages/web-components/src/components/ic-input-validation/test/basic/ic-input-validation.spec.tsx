import { newSpecPage } from "@stencil/core/testing";
import { InputValidation } from "../../ic-input-validation";
import { waitForTimeout } from "../../../../testspec.setup";

describe("ic-input-validation", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [InputValidation],
      html: `<ic-input-validation for="test-id" message="validation message"></ic-input-validation>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with success status", async () => {
    const page = await newSpecPage({
      components: [InputValidation],
      html: `<ic-input-validation for="test-id" message="validation message" status="success"></ic-input-validation>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with warning status", async () => {
    const page = await newSpecPage({
      components: [InputValidation],
      html: `<ic-input-validation for="test-id" message="validation message" status="warning"></ic-input-validation>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with error status", async () => {
    const page = await newSpecPage({
      components: [InputValidation],
      html: `<ic-input-validation for="test-id" message="validation message" status="error"></ic-input-validation>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with full width", async () => {
    const page = await newSpecPage({
      components: [InputValidation],
      html: `<ic-input-validation for="test-id" message="validation message" full-width=true></ic-input-validation>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should update the text content of the aria-live region", async () => {
    const page = await newSpecPage({
      components: [InputValidation],
      html: `<ic-input-validation for="test-id" message="validation message"></ic-input-validation>`,
    });

    page.rootInstance.message = "new validation message";

    await waitForTimeout(200);
    expect(page.root).toMatchSnapshot();
  });
});
