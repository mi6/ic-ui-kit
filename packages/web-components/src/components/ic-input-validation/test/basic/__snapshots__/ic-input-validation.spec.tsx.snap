// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-input-validation should render 1`] = `
<ic-input-validation for="test-id" message="validation message">
  <!---->
  <ic-typography class="statustext" variant="caption">
    <span id="test-id-validation-text">
      validation message
    </span>
    <span aria-live="polite" class="sr-only">
      ​
    </span>
  </ic-typography>
</ic-input-validation>
`;

exports[`ic-input-validation should render with error status 1`] = `
<ic-input-validation class="ic-input-validation-error" for="test-id" message="validation message" status="error">
  <!---->
  <span class="icon-error status-icon">
    svg
  </span>
  <ic-typography class="statustext" variant="caption">
    <span id="test-id-validation-text">
      validation message
    </span>
    <span aria-live="polite" class="sr-only">
      ​
    </span>
  </ic-typography>
</ic-input-validation>
`;

exports[`ic-input-validation should render with full width 1`] = `
<ic-input-validation class="ic-input-validation-full-width" for="test-id" full-width="true" message="validation message">
  <!---->
  <ic-typography class="statustext" variant="caption">
    <span id="test-id-validation-text">
      validation message
    </span>
    <span aria-live="polite" class="sr-only">
      ​
    </span>
  </ic-typography>
</ic-input-validation>
`;

exports[`ic-input-validation should render with success status 1`] = `
<ic-input-validation class="ic-input-validation-success" for="test-id" message="validation message" status="success">
  <!---->
  <span class="icon-success status-icon">
    svg
  </span>
  <ic-typography class="statustext" variant="caption">
    <span id="test-id-validation-text">
      validation message
    </span>
    <span aria-live="polite" class="sr-only">
      ​
    </span>
  </ic-typography>
</ic-input-validation>
`;

exports[`ic-input-validation should render with warning status 1`] = `
<ic-input-validation class="ic-input-validation-warning" for="test-id" message="validation message" status="warning">
  <!---->
  <span class="icon-warning status-icon">
    svg
  </span>
  <ic-typography class="statustext" variant="caption">
    <span id="test-id-validation-text">
      validation message
    </span>
    <span aria-live="polite" class="sr-only">
      ​
    </span>
  </ic-typography>
</ic-input-validation>
`;

exports[`ic-input-validation should update the text content of the aria-live region 1`] = `
<ic-input-validation for="test-id" message="validation message">
  <!---->
  <ic-typography class="statustext" variant="caption">
    <span id="test-id-validation-text">
      validation message
    </span>
    <span aria-live="polite" class="sr-only">
      new validation message
    </span>
  </ic-typography>
</ic-input-validation>
`;
